code/controllers/subsystem/fluids.dm

```diff
@@ -0,0 +1,266 @@
+	var/num_effect_buckets
+	/// The set of buckets containing fluid nodes to process effects for.
+	var/list/effect_carousel
+	/// The index of the currently processing bucket on the effect carousel.
+	var/effect_bucket_index
+	/// The set of fluid nodes we are currently processing effects for.
+	var/list/currently_processing
+	/// Whether the subsystem has resumed processing fluid effects.
+	var/resumed_effect_processing
+
+/datum/controller/subsystem/fluids/Initialize(start_timeofday)
+	initialize_waits()
+	if (case & SS_PROCESSES_SPREADING)
+		initialize_spread_handling()
+	if (case & SS_PROCESSES_EFFECTS)
+		initialize_effect_handling()
+	return ..()
+
+/**
+ * Initializes the subsystem waits.
+ *
+ * Ensures that the subsystem's fire wait evenly splits the spread and effect waits.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_waits()
+	case = NONE
+	if (spread_wait > 0)
+		case |= SS_PROCESSES_SPREADING
+	if (effect_wait > 0)
+		case |= SS_PROCESSES_EFFECTS
+
+	switch (case)
+		if (SS_PROCESSES_SPREADING|SS_PROCESSES_EFFECTS)
+			wait = Gcd(wait, Gcd(spread_wait, effect_wait))
+		if (SS_PROCESSES_SPREADING)
+			if (spread_wait > wait)
+				spread_wait = wait * round(spread_wait / wait, 1)
+			else
+				wait = spread_wait
+			effect_wait = wait
+		if (SS_PROCESSES_EFFECTS)
+			if (effect_wait > wait)
+				effect_wait = wait * round(effect_wait / wait, 1)
+			else
+				wait = effect_wait
+			spread_wait = wait // It needs to be _something_ to prevent runtimes.
+			stack_trace("[src] is a fluid subsystem that does not process fluid spread. Just make it a regular subsystem please.")
+		else
+			spread_wait = wait
+			flags |= SS_NO_FIRE
+			stack_trace("[src] is a fluid subsystem without any purpose. Cull it soon please.")
+
+/**
+ * Initializes the carousel used to process fluid spreading.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_spread_handling()
+	num_spread_buckets = round(spread_wait / wait)
+	spread_wait = wait * num_spread_buckets
+
+	spread_carousel = list()
+	spread_carousel.len = num_spread_buckets
+	for(var/i in 1 to num_spread_buckets)
+		spread_carousel[i] = list()
+	currently_spreading = list()
+	spread_bucket_index = 1
+
+/**
+ * Initializes the carousel used to process fluid effects.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_effect_handling()
+	num_effect_buckets = round(effect_wait / wait)
+	effect_wait = wait * num_effect_buckets
+
+	effect_carousel = list()
+	effect_carousel.len = num_effect_buckets
+	for(var/i in 1 to num_effect_buckets)
+		effect_carousel[i] = list()
+	currently_processing = list()
+	effect_bucket_index = 1
+
+
+/datum/controller/subsystem/fluids/fire(resumed)
+	var/delta_time
+	var/cached_bucket_index
+	var/list/obj/effect/particle_effect/fluid/currentrun
+	MC_SPLIT_TICK_INIT(2)
+
+	MC_SPLIT_TICK // Start processing fluid spread:
+	if(!resumed_spreading)
+		spread_bucket_index = (spread_bucket_index % num_spread_buckets) + 1
```

Comment:
you mind making this a define, like `WRAP()`? `WRAP_UP()` maybe?

Reply:
Sure.